### PR TITLE
fix: Clears old keychain if the configuration changes

### PR DESCRIPTION
**Pull request by royjit**
*Issue #, if available:*

*Description of changes:*

*Check points:*

- [x] Added new tests to cover change, if needed
- [x] All unit tests pass
- [x] All integration tests pass
- [x] Updated CHANGELOG.md
- [] ~Documentation update for the change if required~
- [x] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by harsh62 — APPROVED** [review]

**Comment by palpatim** [comment]
> Could not add test since this require re-configuring AWSMobileClient. This is not possible at runtime

I'm not comfortable with leaving this unautomated. Let's have an out-of-band process populate they keychain with various states so we can test the new code behavior. Off the top of my head:

* Empty: no creds
* Populated: same namespace
* Populated: different namespace

Plus we need to make sure we test different configurations:
* User Pool: yes, Identity Pool: yes
* User Pool: yes, Identity Pool: no
* User Pool: no, Identity Pool: yes (e.g., social federation only)
* User Pool: no, Identity Pool: no (If valid--I'm thinking like an API key configuration--do we support that with AWSMC?)

It means a bit of extra work to write some utility methods, and it'll be slightly brittle since it it requires the test to know the internal storage details of the keychain. But I'd rather have a test that isn't quite as resilient to changes than have this important feature not have sufficient automated tests.

**Comment by royjit** [comment]
# Testing update 1

## 1. App without the fix , signIn then install app with the fix

Expected behavior: The app should work as before, the effect will take place only on the next configuration change of the app.

1.1 Instal with same configuration
Expected: User should be still be signed In
Testing: Success

1.2 Instal with different configuration for CUP and CIDP
Expected: User should be still be signed In
Testing: Success

1.3  Instal with same configuration for CUP and no CIDP
Expected: User should be still be signed In
Testing: Success

1.4  Instal with same configuration for CIDP and no CUP
Expected: User should be still be signed In
Testing: Crashes. This behavior is same as the existing implementation. Not fixed with this bug, refer this issue to track.

1.5 no configuration for CIDP and CUP
Expected: User should be still be signed In
Testing:  Crashes. This behavior is same as the existing implementation. Not fixed with this bug, refer this issue to track.

**Comment by royjit** [comment]
#Testing 2
## 2. App with the fix, signIn then 

2.1 Instal with same configuration
Expected: User should be still be signed In
Testing: Success

2.2 Instal with different configuration for CUP and CIDP
Expected: User should be signed out
Testing: Success


2.3 Instal with same configuration for CUP and no CIDP
Expected: User should be signed out
Testing: Success

2.4 Instal with same configuration for CIDP and no CUP
Expected: User should be signed out
Testing: Success

2.5 Instal with no configuration for CIDP and CUP
Expected: User should be signed out
Testing:  Success

File: AWSAuthSDK/Tests/AWSMobileClientUnitTestWithApp/AWSMobileClientUnitTestWithApp.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
+//
```
Comment (harsh62) [inline AWSAuthSDK/Tests/AWSMobileClientUnitTestWithApp/AWSMobileClientUnitTestWithApp.swift:1]: This file has not tests.. Remove it?

File: AWSAuthSDK/Tests/AWSMobileClientUnitTestWithApp/NameSpaceWithNoPreviousInstallTests.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
+//
```
Comment (harsh62) [inline AWSAuthSDK/Tests/AWSMobileClientUnitTestWithApp/NameSpaceWithNoPreviousInstallTests.swift:1]: Is this the format of the file headers we are going with? Or should it be
```
//
// Copyright Amazon.com Inc. or its affiliates.
// All Rights Reserved.
//
// SPDX-License-Identifier: Apache-2.0
//

```

**Comment by royjit** [comment]
The following scenario needs to be addressed:

1. Start an app with only CIDP configured
2. Check the guest identity ID, let it be ID_1
3. Now add Cognito User Pool 
4. Register a new user and sign In
5. Check the identity ID
6. The id should still be ID_1

**Review by harsh62 — DISMISSED** [review]

**Comment by royjit** [comment]
Tested again all the above test cases and working fine.

**Review by harsh62 — APPROVED** [review]

**Pull request merged**